cpp/include/cudf/utilities/error.hpp

```diff
@@ -99,37 +100,80 @@ struct fatal_cuda_error : public cuda_error {
  * @brief Macro for checking (pre-)conditions that throws an exception when
  * a condition is violated.
  *
- * Example usage:
+ * Defaults to throwing `cudf::logic_error`, but a custom exception may also be
+ * specified.
  *
- * @code
- * CUDF_EXPECTS(lhs->dtype == rhs->dtype, "Column type mismatch");
- * @endcode
+ * Example usage:
+ * ```
+ * // throws cudf::logic_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected null pointer");
  *
- * @param[in] cond Expression that evaluates to true or false
- * @param[in] reason String literal description of the reason that cond is
- * expected to be true
- * @throw cudf::logic_error if the condition evaluates to false.
+ * // throws std::runtime_error
+ * CUDF_EXPECTS(p != nullptr, "Unexpected nullptr", std::runtime_error);
+ * ```
+ * @param ... This macro accepts either two or three arguments:
+ *   - The first argument must be an expression that evaluates to true or
+ *     false, and is the condition being checked.
+ *   - The second argument is a string literal used to construct the `what` of
+ *     the exception.
+ *   - When given, the third argument is the exception to be thrown. When not
+ *     specified, defaults to `cudf::logic_error`.
+ * @throw `_exception_type` if the condition evaluates to 0 (false).
  */
-#define CUDF_EXPECTS(cond, reason)                                  \
-  (!!(cond)) ? static_cast<void>(0)                                 \
-             : throw cudf::logic_error("cuDF failure at: " __FILE__ \
-                                       ":" CUDF_STRINGIFY(__LINE__) ": " reason)
+#define CUDF_EXPECTS(...)                                             \
+  GET_CUDF_EXPECTS_MACRO(__VA_ARGS__, CUDF_EXPECTS_3, CUDF_EXPECTS_2) \
+  (__VA_ARGS__)
+
+/// @cond
+
+#define GET_CUDF_EXPECTS_MACRO(_1, _2, _3, NAME, ...) NAME
+
+#define CUDF_EXPECTS_3(_condition, _reason, _exception_type)                        \
+  do {                                                                              \
+    static_assert(std::is_base_of_v<std::exception, _exception_type>);              \
+    (!!(_condition)) ? static_cast<void>(0)                                         \
```

Comment:
Are there any circumstances under which `!!expr ? true : false` is different from `expr ? true : false` ? I think not because in the former the `!` contextually converts the expression to bool, whereas in the latter, the use as the first operand to `?` contextually converts to bool (IOW, forall a, `static_assert(!!a == bool(a), "Oh noes")` should succeed if conversion to bool is possible, I think)

Reply:
The `!!` logic isn't necessary here. This is probably my fault from the original version I wrote.

The `!!` logic is useful when you're calling an API that takes an `int` that acts like a `bool` instead of a `bool`. e.g., some gcc built-ins work this way.

tl;dr this mechanism is useful in niche use cases, but not necessary here.

Reply:
Right, this is for doing the equivalent of `bool(x)` in Python AFAICT. I can remove it.